tactics/tactics.ml

```diff
@@ -118,8 +118,12 @@ let introduction id =
     in
     let open Context.Named.Declaration in
     match EConstr.kind sigma concl with
-    | Prod (id0, t, b) -> unsafe_intro env (LocalAssum ({id0 with binder_name=id}, t)) b
-    | LetIn (id0, c, t, b) -> unsafe_intro env (LocalDef ({id0 with binder_name=id}, c, t)) b
+    | Prod (id0, t, b) ->
```

Comment:
This is wrong. The goal is supposed to be well-typed, and if you have to do this it means it's not the case.

Reply:
Hi @ppedrot, what do you mean by a goal is supposed to be well-typed? It _is_ the case that relevance marks may be wrong in a goal, due to relevance marks being under-approximated in unification. Here is an example:
```
Axiom P : SProp.
Axiom f : P -> P.
Goal forall H, H = f H.
intro; reflexivity.
(* fails: H computed the mark on an evar initially in `Type` which
    only later was instantiated with an SProp w/o the mark being updated *)
```
Comparitively, this works:
```
Axiom P : SProp.
Axiom f : P -> P.
Goal forall H : P, H = f H.
intro; reflexivity.
```
```

Reply:
Relevance marks are part of the typing annotations of the term. The fact they are not the expected ones simply means that the type of the goal is ill-typed, just as if you wrote a garbage term for the type of the forall binder. We can't write a system where we defensively have to assume that a term is well-typed up to relevance annotations, this is completely nonsensical and will bite us in the long run.

> It is the case that relevance marks may be wrong in a goal

This is the issue that must be fixed, not some workaround when problem has already occurred. Otherwise we'll literally have to adapt all tactics to recompute relevance mark on the fly.

Reply:
> Relevance marks are part of the typing annotations of the term. The fact they are not the expected ones simply means that the type of the goal is ill-typed, just as if you wrote a garbage term for the type of the forall binder. We can't write a system where we defensively have to assume that a term is well-typed up to relevance annotations, this is completely nonsensical and will bite us in the long run.
> 
> > It is the case that relevance marks may be wrong in a goal
> 
> This is the issue that must be fixed, not some workaround when problem has already occurred. Otherwise we'll literally have to adapt all tactics to recompute relevance mark on the fly.

You're right. This PR is somehow a "hot fix".

It is not clear to me, however, if there is another long-run solution than introducing a new kind of mark. E.g. a mark telling that I'm the mark of an evar, waiting for the evar to be instantiated to know what I am.

[We could actually go far in this direction... that when a conversion problem in a type evar of unknown sort is not solvable, we could let open that this type may eventually land in `SProp` and not fail yet.]